modules/scala3-graal/src/main/scala/scala/cli/graal/BytecodeProcessor.scala

```diff
@@ -0,0 +1,253 @@
+  def put(entry: os.RelPath, bytes: Array[Byte]): ClassPathEntry
+}
+
+sealed trait ClassPathEntry {
+  def nioPath: Path = path.toNIO
+  def path: os.Path
+  def modified = true
+}
+case class Unmodified(path: os.Path) extends ClassPathEntry {
+  override def modified: Boolean = false
+}
+case class Processed(path: os.Path, original: os.Path, cache: JarCache) extends ClassPathEntry
+case class CreatedEntry(path: os.Path)                                  extends ClassPathEntry
+
+case class PathingJar(jar: ClassPathEntry, entries: Seq[ClassPathEntry]) extends ClassPathEntry {
+  override def path: os.Path = jar.path
+}
+
+object TempCache extends JarCache {
+  override def cache(path: os.Path)(processPath: os.Path => ClassPathEntry): ClassPathEntry =
+    processPath(
+      if (os.isDir(path)) os.temp.dir(prefix = path.last)
+      else os.temp(prefix = path.baseName, suffix = path.ext)
+    )
+
+  override def put(entry: os.RelPath, content: Array[Byte]): ClassPathEntry = {
+    val path = os.temp(prefix = entry.baseName, suffix = entry.ext)
+    os.write(path, content, createFolders = true)
+    CreatedEntry(path)
+  }
+
+}
+
+object BytecodeProcessor {
+
+  def toClean(classpath: Seq[ClassPathEntry]): Seq[os.Path] = classpath.flatMap {
+    case Processed(path, _, TempCache) => Seq(path)
+    case PathingJar(path, entries)     => toClean(path +: entries)
+    case _                             => Nil
+  }
+
+  def processPathingJar(pathingJar: String, cache: JarCache): Seq[ClassPathEntry] = {
+    val originalJar = toPath(pathingJar)
+    val jarFile     = new JarFile(originalJar.toIO)
+    try {
+      val cp = jarFile.getManifest().getMainAttributes().getValue(Attributes.Name.CLASS_PATH)
+      if (cp != null && cp.nonEmpty) {
+        // paths in pathing jars are spectated by spaces
+        val entries     = cp.split(" +").toSeq
+        val processedCp = processClasspathEntries(entries, cache)
+        val dest        = os.temp(suffix = ".jar")
+        val outStream   = Files.newOutputStream(dest.toNIO, StandardOpenOption.CREATE)
+        try {
+          val stringCp = processedCp.map(_.path.toNIO).mkString(" ")
+          val manifest = new Manifest(jarFile.getManifest())
+          manifest.getMainAttributes().put(Attributes.Name.CLASS_PATH, stringCp)
+          val outjar = new JarOutputStream(outStream, manifest)
+          outjar.close()
+          dest.toNIO.toString()
+          Seq(PathingJar(Processed(dest, originalJar, TempCache), processedCp))
+        }
+        finally outStream.close()
+      }
+      else processClasspathEntries(Seq(pathingJar), cache)
+    }
+    finally jarFile.close()
+  }
+
+  def processClasspath(classpath: String, cache: JarCache = TempCache): Seq[ClassPathEntry] = {
+    val cp = classpath.split(File.pathSeparator)
+    if (cp.size == 1 && cp.head.endsWith(".jar"))
+      processPathingJar(cp.head, cache)
+    else
+      processClasspathEntries(cp.toSeq, cache)
```

Comment:
NIT Match so that we don't call `head`?
```suggestion
    cp match {
      case Array(jar) if jar.endsWith(".jar") => processPathingJar(jar, cache)
      case _ => processClasspathEntries(cp.toSeq, cache)
    }
```